enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,444 @@
+
+Also, it doesn't cover the changes required for OpenStack platform to support Failure Domains and is covered in another [enhancement](https://github.com/openshift/enhancements/pull/1167).
+
+## Proposal
+
+We will automate additional capabilities in cluster network infrastructure, which are already handled by some other platforms
+infrastructure services.
+
+These capabilities include:
+- Highly available and dynamically routed load-balanced api access
+- Highly available and dynamically routed ingress access
+
+The assets needed for these capabilities implementation are rendered by the [MCO](https://github.com/openshift/machine-config-operator) and
+the [baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg) is used significantly within the manifests and templates of
+the machine-config-operator project.
+There are also changes in the API and installer projects, which are further described in sections below.
+
+### Workflow Description
+
+The installer accepts an LB type (default remains what it does today), which will be set into the Cluster Infrastructure Object. Then MCO will read the LB type and if it's BGP it'll not deploy Keepalived/HAproxy but FRR with a generated configuration.
+
+### Spine-Leaf architecture
+
+![image](on-prem-bgp-networking-spine-leaf.png)
+
+This architecture is a common design for data centers that have a spine-leaf architecture.
+The OpenShift nodes are connected to the leafs, which are connected to the spine.
+In our proposal, the control plane will peer with the leafs, and the leafs will peer with the spine, so the VIPs
+will be routable across the network.
+
+### API
+
+In conjunction to the work that has been (and is being) done on Failure Domains, we will propose 
+a new [API](https://github.com/openshift/api/pull/1321) to configure Load Balancers, so we will be able to:
+- Change the load-balancer type (default remains VRRP), to be able to enable BGP.
+- If BGP is enabled, allow to provide a list of BGP speakers and their configurations.
+- The Machine Config Operator (MCO) will monitor this setting and generate the required configuration.
+
+### Installer
+
+A new [installer](https://github.com/openshift/installer/pull/6440) PR has been proposed, so it'll be possible to:
+- Populate the load-balancer type in the Cluster Infrastructure object, so it can be consumed later (e.g. by MCO).
+- Change the load-balancer type (default remains VRRP) all from `install-config.yaml`.
+
+TODO - we need to clarify (comments welcome):
+On some platforms, the VIPs could either be floating IPs or just part of the machine subnets. In this scenario, the
+VIPs are floatable between failure domains, so they aren't part of the machine subnets. Do we want a new interface
+to define the VIP networks/subnets? In some platforms (e.g. OpenStack), we'll need to inform IPAM that these IPs
+need to be reserved so they aren't taken by something else in the infrastructure. An alternative is to not
```

Comment:
Does OpenShift do any IPAM? How would we inform IPAM? Is it not up to the admin configuring the BGP to sort out the IPAM side?

Reply:
> Does OpenShift do any IPAM?
I don't think OpenShift does IPAM for the control plane network, and it's up to the OpenShift deployer to manage the machine networks.
With IPI, it's done via the installer (terraform).

> How would we inform IPAM?
In OpenStack, we would create the ports in Neutron, in "DOWN" status (so OVN won't actually create anything) so they'll be "reserved" and nobody can create an Neutron port (in UP) with these IPs.

> Is it not up to the admin configuring the BGP to sort out the IPAM side?
yeah that's my question here...

Reply:
There's an ongoing vSphere project about static IPs where we are designing it so that the IPAM is handled by the user, so I think that's what we will do here too

Reply:
I just saw https://github.com/openshift/enhancements/pull/1267 - gotcha. I'll include that here.